tests/BrowserTest.php

```diff
@@ -503,4 +503,114 @@ public function testCancelGetRequestShouldCancelUnderlyingSocketConnection()
         $promise = $this->browser->get('http://example.com/');
         $promise->cancel();
     }
+
+    public function testWithElseHeader()
+    {
+        $this->browser = $this->browser->withHeader('User-Agent', 'ACMC');
+
+        $that = $this;
+        $this->sender->expects($this->once())->method('send')->with($this->callback(function (RequestInterface $request) use ($that) {
+            $that->assertEquals(array('ACMC'), $request->getHeader('User-Agent'));
+            return true;
+        }))->willReturn(new Promise(function () { }));
+
+        $this->browser->get('http://example.com/');
+    }
+
+    public function testWithHeaderShouldOverwriteExistingHeader()
+    {
+        $this->browser = $this->browser->withHeader('User-Agent', 'ACMC'); //should be overwritten
+        $this->browser = $this->browser->withHeader('user-agent', 'ABC'); //should be the user-agent
+
+        $that = $this;
+        $this->sender->expects($this->once())->method('send')->with($this->callback(function (RequestInterface $request) use ($that) {
+            $that->assertEquals(array('ABC'), $request->getHeader('UsEr-AgEnT'));
+            return true;
+        }))->willReturn(new Promise(function () { }));
+
+        $this->browser->get('http://example.com/');
+    }
+
+    public function testWithHeadersShouldBeMergedCorrectlyWithDefaultHeaders()
+    {
+        $this->browser = $this->browser->withHeader('User-Agent', 'ACMC');
+
+        $that = $this;
+        $this->sender->expects($this->once())->method('send')->with($this->callback(function (RequestInterface $request) use ($that) {
+            $that->assertEquals(array('ABC'), $request->getHeader('UsEr-AgEnT'));
+            return true;
+        }))->willReturn(new Promise(function () { }));
+
+        $this->browser->get('http://example.com/', array('user-Agent' => 'ABC')); //should win
+    }
+
+    public function testWithMultipleHeadersShouldBeMergedCorrectlyWithMultipleDefaultHeaders()
```

Comment:
The test name is rather vague. Does this test anything that's not already covered by the other tests?

Reply:
This method tests multiply headers. In the other tests i only tested one header at the time. That was complely fine, until i tested it with multiply headers passd into e.g. `$browser->get('myurl', $headers)`. I found out that this will cause a bug, which will allways overwrite a header. Thanks to @SimonFrings, who told me to write such a test, this bug was found.

Reply:
Alright, sounds good :+1: